File: lib/MySQL.lua
local Store = {}

local function addStore(query, cb)
	assert(type(query) == 'string', 'The SQL Query must be a string')
	local store = #Store+1
	Store[store] = query
	if cb then cb(store) else return store end
end

local MySQL = {
	Sync = { store = addStore },
	Async = { store = addStore },

	ready = function(cb)
		CreateThread(function()
			repeat
				Wait(50)
			until GetResourceState('oxmysql') == 'started'
			cb()
		end)
	end
}

local type = type

local function safeArgs(query, parameters, cb, transaction)
	local queryType = type(query)

	if queryType == 'number' then
		query = Store[query]
	elseif transaction then
		if queryType ~= 'table' then
			error(("First argument expected table, received '%s'"):format(query))
		end
	elseif queryType ~= 'string' then
		error(("First argument expected string, received '%s'"):format(query))
	end

	if parameters then
		local paramType = type(parameters)

		if paramType ~= 'table' and paramType ~= 'function' then
			error(("Second argument expected table or function, received '%s'"):format(parameters))
		end

		if paramType == 'function' or parameters.__cfx_functionReference then
			cb = parameters
			parameters = nil
		end
	end

	if cb and parameters then
		local cbType = type(cb)

		if cbType ~= 'function' and (cbType == 'table' and not cb.__cfx_functionReference) then
			error(("Third argument expected function, received '%s'"):format(cb))
		end
	end

	return query, parameters, cb
end

local promise = promise
local oxmysql = exports.oxmysql
local Await = Citizen.Await
local GetCurrentResourceName = GetCurrentResourceName()

local function await(fn, query, parameters)
	local p = promise.new()
	fn(nil, query, parameters, function(result, error)
		if error then
			return p:reject(error)
		end

		p:resolve(result)
	end, GetCurrentResourceName, true)
	return Await(p)
end

setmetatable(MySQL, {
	__index = function(self, method)
		local state = GetResourceState('oxmysql')
		if state == 'started' or state == 'starting' then
			self[method] = setmetatable({}, {

				__call = function(_, query, parameters, cb)
					query, parameters, cb = safeArgs(query, parameters, cb, method == 'transaction')
					return oxmysql[method](nil, query, parameters, cb, GetCurrentResourceName, false)
				end,

				__index = function(_, index)
					assert(index == 'await', ('unable to index MySQL.%s.%s, expected .await'):format(method, index))
					self[method].await = function(query, parameters)
						return await(oxmysql[method], safeArgs(query, parameters, nil, method == 'transaction'))
					end
					return self[method].await
				end
			})

			return self[method]
		else
			error(('^1oxmysql resource state is %s - unable to trigger exports.oxmysql:%s^0'):format(state, method), 0)
		end
	end
})

local alias = {
	fetchAll = 'query',
	fetchScalar = 'scalar',
	fetchSingle = 'single',
	insert = 'insert',
	execute = 'update',
	transaction = 'transaction',
	prepare = 'prepare'
}

local alias_mt = {
	__index = function(self, key)
		if alias[key] then
			MySQL.Async[key] = MySQL[alias[key]]
			MySQL.Sync[key] = MySQL[alias[key]].await
			alias[key] = nil
			return self[key]
		end
	end
}

setmetatable(MySQL.Async, alias_mt)
setmetatable(MySQL.Sync, alias_mt)

_ENV.MySQL = MySQL

File: lib/MySQL.ts
type Query = string | number;
type Params = Record<string, unknown> | unknown[] | Function;
type Callback<T> = (result: T | null) => void;

type Transaction = Object[]

interface Result {
    [column: string | number]: any;
    affectedRows?: number;
    fieldCount?: number;
    info?: string;
    insertId?: number;
    serverStatus?: number;
    warningStatus?: number;
    changedRows?: number;
}

interface Row {
    [column: string | number]: unknown;
}

interface OxMySQL {
  store: (query: string) => void;
  ready: (callback: () => void) => void;
  query: <T = Result | null> (query: Query, params?: Params | Callback<T>, cb?: Callback<T>) => Promise<T>
  single: <T = Row | null> (query: Query, params?: Params | Callback<Exclude<T, []>>, cb?: Callback<Exclude<T, []>>) => Promise<Exclude<T, []>>
  scalar: <T = unknown | null> (query: Query, params?: Params | Callback<Exclude<T, []>>, cb?: Callback<Exclude<T, []>>) => Promise<Exclude<T, []>>
  update: <T = number | null> (query: Query, params?: Params | Callback<T>, cb?: Callback<T>) => Promise<T>
  insert: <T = number | null> (query: Query, params?: Params | Callback<T>, cb?: Callback<T>) => Promise<T>
  prepare: <T = any> (query: Query, params?: Params | Callback<T>, cb?: Callback<T>) => Promise<T>
  transaction: (query: Transaction, params?: Params | Callback<boolean>, cb?: Callback<boolean>) => Promise<boolean>
}

const QueryStore: string[] = [];

function assert(condition: boolean, message: string) {
  if (!condition) throw new TypeError(message);
}

const safeArgs = (
  query: Query | Transaction,
  params?: any,
  cb?: Function,
  transaction?: true
) => {
  if (typeof query === "number") query = QueryStore[query];

  if (transaction) {
    assert(
      typeof query === "object",
      `First argument expected object, recieved ${typeof query}`
    );
  } else {
    assert(
      typeof query === "string",
      `First argument expected string, received ${typeof query}`
    );
  }

  if (params) {
    const paramType = typeof params;
    assert(
      paramType === "object" || paramType === "function",
      `Second argument expected object or function, received ${paramType}`
    );

    if (!cb && paramType === "function") {
        cb = params
        params = undefined
    }
  }

  if (cb !== undefined)
    assert(
      typeof cb === "function",
      `Third argument expected function, received ${typeof cb}`
    );

  return [query, params, cb];
};

const exp = global.exports.oxmysql;
const currentResourceName = GetCurrentResourceName()

function execute(method: string, query: Query | Transaction, params?: Params) {
    return new Promise((resolve, reject) => {
        exp[method](query, params, (result, error) => {
            if (error) return reject(error)
            resolve(result)
        }, currentResourceName, true)
    }) as any
}

export const oxmysql: OxMySQL = {
  store(query) {
    assert(
      typeof query !== "string",
      `Query expects a string, received ${typeof query}`
    );

    return QueryStore.push(query);
  },
  ready(callback) {
    setImmediate(async () => {
      while (GetResourceState("oxmysql") !== "started")
        await new Promise((resolve) => setTimeout(resolve, 50));
      callback();
    });
  },
  async query(query, params, cb) {
    [query, params, cb] = safeArgs(query, params, cb)
    const result = await execute('query', query, params)
    return cb ? cb(result) : result
  },
  async single(query, params, cb) {
    [query, params, cb] = safeArgs(query, params, cb)
    const result = await execute('single', query, params)
    return cb ? cb(result) : result
  },
  async scalar(query, params, cb) {
    [query, params, cb] = safeArgs(query, params, cb)
    const result = await execute('scalar', query, params)
    return cb ? cb(result) : result
  },
  async update(query, params, cb) {
    [query, params, cb] = safeArgs(query, params, cb)
    const result = await execute('update', query, params)
    return cb ? cb(result) : result
  },
  async insert(query, params, cb) {
    [query, params, cb] = safeArgs(query, params, cb)
    const result = await execute('insert', query, params)
    return cb ? cb(result) : result
  },
  async prepare(query, params, cb) {
    [query, params, cb] = safeArgs(query, params, cb)
    const result = await execute('prepare', query, params)
    return cb ? cb(result) : result
  },
  async transaction(query, params, cb) {
    [query, params, cb] = safeArgs(query, params, cb, true)
    const result = await execute('transaction', query, params)
    return cb ? cb(result) : result
  },
};
